zkvm/docs/zkvm-spec.md

```diff
@@ -828,6 +787,54 @@ V == v·B + 0·B2
 1. Prover shows `v`.
 2. Verifier checks equality `V == v·B`.
 
+### Taproot
+
+Taproot provides efficient and privacy-preserving storage of smart contracts. It is based on [a proposal by Gregory Maxwell](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-January/015614.html). Multi-party blockchain contracts typically have a top-level "success clause" (all parties agree and sign) or "alternative clauses" to let a party exit the contract based on pre-determined constraints. This has three significant features.
+
+1. The alternative clauses `{C_i}`, each of which is a [program](#program), and their corresponding blinding factors are stored as [blinded programs](#blinded-program) and compressed in a [Merkle tree](#merkle-binary-tree) with root `M`.
+2. A signing key `X` and the Merkle root `M` (from 1) are committed to a single signing key `P` using a hash function `h1`, such that `P = X + h1(X, M)`. This makes signing for `P` possible if parties want to sign for `X` and avoids revealing the alternative clauses.
+3. Calling a program will check a [call proof](#call-proof) to verify the program's inclusion in the Taproot tree before executing the program.
+
+The commitment scheme is defined using the [transcript](#transcript) protocol by committing (1) `X`, the signing key, as a compressed 32-byte point, and (2) `M`, the root of the Merkle tree constructed from the `n` programs and (3) squeezing a scalar bound to all `n` programs.
+
+```
+// given Merkle root M and signing key X
+t = Transcript("ZkVM.taproot")
+t.commit_bytes(b"key", X); // Compressed Ristretto
+t.commit_bytes(b"merkle", M); // [u8; 32]
+t.challenge_scalar(b"h")
+```
+
+Typically, to select a program to run from a leaf of the Merkle tree, we must navigate branches from the Merkle root. We tweak this procedure: the Merkle root is constructed from a provided Merkle path. It is then used to verify a relation between the contract predicate and the provided program. A branch is chosen using the [`call`](#call) instruction, which takes a program, [call proof](#call-proof), and contract. The [predicate](#predicate) `P` is read from the contract. The call proof contains the signing key `X`, a list of left-right neighbors, and the Merkle path. The neighbor list, Merkle path, and the provided program are used to reconstruct the Merkle root `M`. If the relation `P = X + h1(X, M)` is verified, then the associated program will be executed.
```

Comment:
>Typically, to select a program to run from a leaf of the Merkle tree, we must navigate branches from the Merkle root. We tweak this procedure: the Merkle root is constructed from a provided Merkle path.

Wdym by "we tweak this procedure"?

Reply:
> then the associated program will be executed.
Associated with what? You say "provided program" already, so we can probably say "the program" simply, since there's no other program. Qualifying it with "associated" or any other adjective makes the reader double-check if there are any other programs to consider.

Reply:
First comment: I wanted to communicate that rather than being given a root and navigating to the leaf from it, we are given the path and leaf and then compute the root for verification. It's very possible that this does not require any additional emphasis at all, so I'm happy to simply take that phrase out.

Second comment: this makes sense, I'm changing it to "the program".